### ISSUE TITLE
Allow `Bump::unallocated` for any `A` that implements `Default`

### ISSUE THREAD

**Issue by bluurryy** [bluurryy/bump-scope#13]
We currently only allow `Global` because we know it can be default constructed without the use of `Default` via `MaybeUninit::uninit().assume_init()`.

**Comment by bluurryy** [comment]
This implies that any `Bump<GUARANTEED_ALLOCATED = false>` must implement `Default` for `A` to do any allocations.

We could have duplicate implementations for `alloc_*` functions for `true` and `false` of `GUARANTEED_ALLOCATED`. But that would mess with docs and generic usage.

With specialization we could do something like this:
```rust
#![feature(specialization)]

trait DefaultOrPanic: Sized {
    fn default_or_panic() -> Self;
}

impl<T> DefaultOrPanic for T {
    default fn default_or_panic() -> Self {
        panic!("type does not implement `Default`")
    }
}

impl<T> DefaultOrPanic for T where T: Default {
    fn default_or_panic() -> Self {
        T::default()
    }
}
```

Where we only call `default_or_panic` when `GUARANTEED_ALLOCATED = false` which can only have been created with `A: Default` so the panicking implementation is unreachable.

**Comment by bluurryy** [comment]
We can express these bounds in stable like this:
```rust
struct Alloc<const GUARANTEED_ALLOCATED: bool, A>(PhantomData<A>);

trait SupportedAlloc<A> {
    // ...
}

impl<A> SupportedAlloc<A> for Alloc<true, A> {
    // ...
}

impl<A> SupportedAlloc<A> for Alloc<false, A> 
where
    A: Default
{
    // ...
}
```

With this line added to the where clause of `Bump(Scope)` alloc impls:
```rust
Alloc<GUARANTEED_ALLOCATED, A>: SupportedAlloc<A>
```

The names are just placeholder, I'm not happy with them.

**Comment by bluurryy** [comment]
Implemented in 97fc797. (In a different way)
